network/builder_test.go

```diff
@@ -45,14 +49,118 @@ func TestSetters(t *testing.T) {
 		t.Fatalf("address is wrong: expected %s but got %s", fmt.Sprintf("tcp://127.0.0.1:%d", port), net.Address)
 	}
 
-	if !bytes.Equal(net.Keys.PrivateKey, keys.PrivateKey) {
+	if !bytes.Equal(net.keys.PrivateKey, keys.PrivateKey) {
 		t.Fatalf("private key is wrong")
 	}
 
-	if !bytes.Equal(net.Keys.PublicKey, keys.PublicKey) {
+	if !bytes.Equal(net.keys.PublicKey, keys.PublicKey) {
 		t.Fatalf("public key is wrong")
 	}
+}
+
+func TestNoKeys(t *testing.T) {
+	t.Parallel()
+
+	builder := NewBuilder()
+	builder.SetKeys(nil)
+	_, err := builder.Build()
+	if err == nil {
+		t.Errorf("Build() = nil, expected %v", ErrNoKeyPair)
+	}
+}
+
+func TestBuilderAddress(t *testing.T) {
+	t.Parallel()
+
+	builder := NewBuilder()
+	builder.SetAddress("")
+	_, err := builder.Build()
+	if err == nil || err != ErrNoAddress {
+		t.Errorf("Build() = %v, expected %v", err, ErrNoAddress)
+	}
+
+	errMissingPort := errors.New("missing port in address")
+	builder.SetAddress("localhost")
+	_, err = builder.Build()
+	if err == nil || err.Error() != errMissingPort.Error() {
+		t.Errorf("Build() = %v, expected %v", err, errMissingPort)
+	}
+}
+
+func TestDuplicatePlugin(t *testing.T) {
+	t.Parallel()
+
+	builder := NewBuilder()
+	_, err := builder.Build()
+	if err != nil {
+		t.Errorf("Build() = %v, expected <nil>", err)
+	}
+	assert.Equal(t, builder.pluginCount, 0, "should have no plugins")
+
+	err = builder.AddPluginWithPriority(1, new(MockPlugin))
+	if err != nil {
+		t.Errorf("AddPluginWithPriority() = %v, expected <nil>", err)
+	}
+
+	err = builder.AddPluginWithPriority(1, new(MockPlugin))
+	if err == nil || err != ErrDuplicatePlugin {
+		t.Errorf("Build() = %v, expected %v", err, ErrDuplicatePlugin)
+	}
+}
+
+func TestConnectionTimeout(t *testing.T) {
+	t.Parallel()
+
+	timeout := 5 * time.Second
+	builder := NewBuilderWithOptions(ConnectionTimeout(timeout))
+	net, err := builder.Build()
+	if err != nil {
+		t.Errorf("Build() = %v, expected <nil>", err)
+	}
+	if net.opts.connectionTimeout != timeout {
+		t.Errorf("connectionTimeout = %v, expected %v", net.opts.connectionTimeout, timeout)
+	}
+}
+
+func TestSignaturePolicy(t *testing.T) {
+	t.Parallel()
+
+	signaturePolicy := ed25519.New()
+	builder := NewBuilderWithOptions(SignaturePolicy(signaturePolicy))
+	net, err := builder.Build()
+	if err != nil {
+		t.Errorf("Build() = %v, expected <nil>", err)
+	}
+	assert.Equal(t, net.opts.signaturePolicy, signaturePolicy, "signature policy given should match found")
+}
+
+func TestHashPolicy(t *testing.T) {
+	t.Parallel()
 
+	hashPolicy := blake2b.New()
+	builder := NewBuilderWithOptions(HashPolicy(hashPolicy))
+	net, err := builder.Build()
+	if err != nil {
+		t.Errorf("Build() = %v, expected <nil>", err)
+	}
```

Comment:
to be consistent, can use 
```
assert.NotEqual(t, nil, err)
```
or require.NotEqual(t, nil, err) from "github.com/stretchr/testify/require"